src/dawn/Optimizer/PassSetNonTempCaches.cpp

```diff
@@ -0,0 +1,325 @@
+#include "dawn/Optimizer/StencilInstantiation.h"
+#include "dawn/SIR/ASTExpr.h"
+#include "dawn/Support/Unreachable.h"
+#include <iostream>
+#include <set>
+#include <vector>
+
+namespace dawn {
+
+struct accessIDaccessMetric {
+  accessIDaccessMetric(int id, int gain) : accessID(id), dataLocalityGain(gain) {}
+  int accessID;
+  int dataLocalityGain;
+  bool operator<(const accessIDaccessMetric& rhs) {
+    return dataLocalityGain < rhs.dataLocalityGain;
+  }
+};
+
+struct nameToImprovementMetric {
+  nameToImprovementMetric(std::string name_, Cache cache_, int metric)
+      : name(name_), cache(cache_), dataLocalityImprovement(metric) {}
+  std::string name;
+  Cache cache;
+  int dataLocalityImprovement;
+};
+
+/// @brief The GlobalFieldCacher class handles the caching for a given Multistage
+/// @param[in, out]  msprt   Pointer to the multistage to handle
+/// @param[in, out]  si      Stencil Instanciation [ISIR] holding all the Stencils
+class GlobalFieldCacher {
+public:
+  GlobalFieldCacher(MultiStage* msptr, StencilInstantiation* si)
+      : multiStagePrt_(msptr), instantiation_(si) {}
+
+  /// @brief entry method for the pass: processes a given multistage and applies all changes
+  /// required
+  void process() {
+    computeOptimalFields();
+    addFillerStages();
+    addFlushStages();
+  }
+
+  std::vector<nameToImprovementMetric> getoriginalNameToCache() { return originalNameToCache_; }
+
+private:
+  /// @brief use the data locality metric to rank the fields in a stencil based on how much they
+  /// would benefit from caching
+  void computeOptimalFields() {
+    auto dataLocality =
+        computeReadWriteAccessesMetricIndividually(instantiation_, *multiStagePrt_, config_);
+
+    for(const auto& stagePtr : multiStagePrt_->getStages()) {
+      for(const Field& field : stagePtr->getFields()) {
+        auto findField = accessIDtoDataLocality_.find(field.AccessID);
+
+        // We already checked this field
+        if(findField != accessIDtoDataLocality_.end())
+          continue;
+
+        // Only ij-cache fields that are not horizontally pointwise and are vertically pointwise
+        if(!(field.Extent.isVerticalPointwise()) || (field.Extent.isHorizontalPointwise()))
+          continue;
+
+        // This is caching non-temporary fields
+        if(instantiation_->isTemporaryField(field.AccessID))
+          continue;
+
+        int cachedreadwrites = dataLocality.find(field.AccessID)->second.total();
+
+        // We check how much the cache-filling costs: 1 Read and if we fill from Memory
+        if(checkReadBeforeWrite(field.AccessID))
+          cachedreadwrites--;
+        // We need one more write to flush the cache back to the variable
+        if(!checkReadOnlyAccess(field.AccessID))
+          cachedreadwrites--;
+
+        if(cachedreadwrites > 0)
+          accessIDtoDataLocality_.emplace(field.AccessID, cachedreadwrites);
+      }
+    }
+
+    for(auto& performanceMesure : accessIDtoDataLocality_) {
```

Comment:
What does `performanceMesure` mean here? Isn't this just a pair of AccessIDs to the value of the metric? How about naming this `AccessIDMetricPair`?